REF: separate indexer utilities from indexing.py

jbrockmendel opened the pull request:
A couple things going on here.

- Take simple, low-dependency, not-used-in-indexing.py functions out of indexing.py and put them in indexers.py.
  - A couple of runtime-defined functions in internals get moved here too.
  - Avoids some runtime imports (i.e. simplify dependency structure)
  - Fixes some incorrect private names
  - Adds typing
- Simplify some indexing.py methods
  - Avoid duplicated validation calls
  - Add typing
  - De-duplicate _get_slice_axis method
  - Move code outside of try/except blocks to clarify what the failure modes are.

Medium-term goals in indexing.py
  - Isolate internals access
  - refactor _setitem_with_indexer (over 300 lines)

jbrockmendel committed:
implement indexers.py

jbrockmendel committed:
add types

jbrockmendel committed:
cleanup

jbrockmendel committed:
separate out helper method; add types, remove redundnat checks

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into clnindexing

jbrockmendel committed:
blackify

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into clnindexing

jbrockmendel committed:
types

jbrockmendel committed:
move indexers

jbrockmendel committed:
organize file

jreback commented on pandas/core/indexers.py:
```diff
@@ -0,0 +1,211 @@
+# Indexer Identification
+
+def is_list_like_indexer(key) -> bool:
+    # allow a list_like, but exclude NamedTuples which can be indexers
```
can u add as proper doc-strings

jbrockmendel replied:
sure

jreback commented on pandas/core/indexers.py:
```diff
@@ -0,0 +1,211 @@
+    return is_list_like(key) and not (isinstance(key, tuple) and type(key) is not tuple)
+
+
+def is_scalar_indexer(indexer, arr_value) -> bool:
```
if possible to type pls do so (may be tricky)

I would type with Any explicity if it’s truly that
so we know this is done (meaning it has been typed)

possible to use reveal_type to figure things out here

jbrockmendel replied:
ATM I've added types for everything I had figured out with sufficient certainty.  I'd rather leave it clearly-unfinished so it gets more attention than put Any if we can be more specific

jreback replied:
can you fix this doc-sring in followup and type as much as possible here

jreback reviewed (changes requested):
just some quick comments
will have a full look later

+1 on refactoring things

jreback commented:
note as a post or pre-cursor to this I would form pandas/core/indexing/ and move indexing.py (and indexers.py) there as well

jbrockmendel commented:
> note as a post or pre-cursor to this I would form pandas/core/indexing/ and move indexing.py (and indexers.py) there as well

Sure.  Obvious preference for post-cursor.

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into clnindexing

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into clnindexing

jbrockmendel committed:
isort

pep8speaks commented:
Hello @jbrockmendel! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`pandas/core/indexers.py`](https://github.com/pandas-dev/pandas/blob/3322f4e85ea66b90f6ddb901dca0b1d73f5e7002/pandas/core/indexers.py):

> [Line 12:1](https://github.com/pandas-dev/pandas/blob/3322f4e85ea66b90f6ddb901dca0b1d73f5e7002/pandas/core/indexers.py#L12): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1

jbrockmendel committed:
lint+isort compat

jbrockmendel committed:
blackify

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into clnindexing

jreback commented on pandas/core/indexing.py:
```diff
@@ -1329,20 +1308,16 @@ def _tuplify(self, loc):
-        else:
-            return self.obj._take(indexer, axis=axis)
+        indexer = self._convert_slice_indexer(slice_obj, axis)
+        assert isinstance(indexer, slice), type(indexer)
```
prob better off to have this assert in _slice

jreback commented on pandas/core/indexing.py:
```diff
@@ -1450,11 +1425,11 @@ class _LocationIndexer(_NDFrameIndexer):
     def __getitem__(self, key):
         if type(key) is tuple:
```
should prob use isinstance here as this is an old idiom

jbrockmendel replied:
tried this, turns out this is intentional to get tuple but exclude NamedTuple.  Added a comment to that effect.  (green)

jreback reviewed (changes requested):
just a couple of doc-strings I think (in the first section), then can merge this

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into clnindexing

jbrockmendel committed:
remove assertion

jbrockmendel committed:
add comment for GH#27259

jbrockmendel committed:
docstring

jbrockmendel commented:
Added one more docstring, left un-docstringed is_scalar_indexer and is_empty_indexer because I'm not totally clear on what those functions are doing (possibly an extra indentation in is_scalar_indexer?  it is _almost_ identical to the last 5 lines of is_empty_indexer)

Added a note about InvalidIndexError being for geopandas.

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into clnindexing

jbrockmendel committed:
revert isinstance to type checks

jreback commented on pandas/core/indexers.py:
```diff
@@ -0,0 +1,225 @@
+# Indexer Validation
+
+
+def check_setitem_lengths(indexer, value, values) -> None:
```
type

jreback commented:
thanks. in followups we need to move / consolidate these things; also a typing pass would be helpful.
